crates/viewer/re_space_view_dataframe/src/dataframe_ui.rs

```diff
@@ -0,0 +1,350 @@
+    batch_and_row_from_row_nr: BTreeMap<u64, (usize, usize)>,
+
+    /// The index of the time column corresponding to the query timeline.
+    time_column_index: Option<usize>,
+
+    /// The index of the time column corresponding the row IDs.
+    row_id_column_index: Option<usize>,
+}
+
+impl RowsDisplayData {
+    fn try_new(
+        row_indices: &Range<u64>,
+        record_batches: Vec<RecordBatch>,
+        schema: &[ColumnDescriptor],
+        query_timeline: &Timeline,
+    ) -> Result<Self, DisplayRecordBatchError> {
+        let display_record_batches = record_batches
+            .into_iter()
+            .map(|record_batch| DisplayRecordBatch::try_new(&record_batch, schema))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let mut batch_and_row_from_row_nr = BTreeMap::new();
+        let mut offset = row_indices.start;
+        for (batch_idx, batch) in display_record_batches.iter().enumerate() {
+            let batch_len = batch.num_rows() as u64;
+            for row_idx in 0..batch_len {
+                batch_and_row_from_row_nr.insert(offset + row_idx, (batch_idx, row_idx as usize));
+            }
+            offset += batch_len;
+        }
+
+        // find the time column
+        let time_column_index = schema
+            .iter()
+            .find_position(|desc| match desc {
+                ColumnDescriptor::Time(time_column_desc) => {
+                    &time_column_desc.timeline == query_timeline
+                }
+                _ => false,
+            })
+            .map(|(pos, _)| pos);
+
+        // find the row id column
+        let row_id_column_index = schema
+            .iter()
+            .find_position(|desc| match desc {
+                ColumnDescriptor::Control(control_column_desc) => {
+                    control_column_desc.component_name
+                        == ComponentName::from("rerun.controls.RowId")
+                }
+                _ => false,
+            })
+            .map(|(pos, _)| pos);
+
+        Ok(Self {
+            display_record_batches,
+            batch_and_row_from_row_nr,
+            time_column_index,
+            row_id_column_index,
+        })
+    }
+}
+
+/// [`egui_table::TableDelegate`] implementation for displaying a [`QueryHandle`] in a table.
+struct DataframeTableDelegate<'a> {
+    ctx: &'a ViewerContext<'a>,
+    query_handle: &'a QueryHandle<'a>,
+    schema: &'a [ColumnDescriptor],
+    header_entity_paths: Vec<Option<EntityPath>>,
+    display_data: anyhow::Result<RowsDisplayData>,
+
+    num_rows: u64,
+}
+
+impl<'a> egui_table::TableDelegate for DataframeTableDelegate<'a> {
+    fn prepare(&mut self, info: &egui_table::PrefetchInfo) {
+        re_tracing::profile_function!();
+
+        let data = RowsDisplayData::try_new(
+            &info.visible_rows,
+            self.query_handle.get(
+                info.visible_rows.start,
+                info.visible_rows.end - info.visible_rows.start,
```

Comment:
If we use `Range` in the API instead

```suggestion
                info.visible_rows,
```